File: app/src/main/java/com/electricdreams/shellshock/feature/items/ItemListActivity.java
package com.electricdreams.shellshock.feature.items;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.electricdreams.shellshock.R;
import com.electricdreams.shellshock.core.model.Item;
import com.electricdreams.shellshock.core.util.ItemManager;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.Currency;
import java.util.List;
import java.util.Locale;

public class ItemListActivity extends AppCompatActivity {

    private ItemManager itemManager;
    private RecyclerView recyclerView;
    private TextView emptyView;
    private ItemAdapter adapter;

    private final ActivityResultLauncher<Intent> addItemLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    refreshItems();
                }
            });

    private final ActivityResultLauncher<String> selectCsvLauncher = registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    processCsvFile(uri);
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_item_list);

        // Set up toolbar
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle(R.string.item_list_title);

        // Initialize views
        recyclerView = findViewById(R.id.items_recycler_view);
        emptyView = findViewById(R.id.empty_view);
        FloatingActionButton fabAddItem = findViewById(R.id.fab_add_item);

        // Get item manager instance
        itemManager = ItemManager.getInstance(this);

        // Set up RecyclerView
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new ItemAdapter(itemManager.getAllItems());
        recyclerView.setAdapter(adapter);

        // Set up empty view
        updateEmptyViewVisibility();

        // Set up add item FAB
        fabAddItem.setOnClickListener(v -> {
            Intent intent = new Intent(ItemListActivity.this, ItemEntryActivity.class);
            addItemLauncher.launch(intent);
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void refreshItems() {
        adapter.updateItems(itemManager.getAllItems());
        updateEmptyViewVisibility();
    }

    private void updateEmptyViewVisibility() {
        if (adapter.getItemCount() == 0) {
            emptyView.setVisibility(View.VISIBLE);
            recyclerView.setVisibility(View.GONE);
        } else {
            emptyView.setVisibility(View.GONE);
            recyclerView.setVisibility(View.VISIBLE);
        }
    }

    private void processCsvFile(Uri uri) {
        // Implement CSV file processing here
        Toast.makeText(this, "Processing CSV file...", Toast.LENGTH_SHORT).show();
        
        // Would need to copy the file to a temp location and then import it
        // For this demo, we'll manually add some sample items instead
        Toast.makeText(this, "CSV import not implemented in this demo", Toast.LENGTH_SHORT).show();
    }

    private class ItemAdapter extends RecyclerView.Adapter<ItemAdapter.ItemViewHolder> {
        
        private List<Item> items;
        
        ItemAdapter(List<Item> items) {
            this.items = items;
        }
        
        void updateItems(List<Item> newItems) {
            this.items = newItems;
            notifyDataSetChanged();
        }

        @NonNull
        @Override
        public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_product, parent, false);
            return new ItemViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull ItemViewHolder holder, int position) {
            Item item = items.get(position);
            holder.bind(item);
        }

        @Override
        public int getItemCount() {
            return items.size();
        }

        class ItemViewHolder extends RecyclerView.ViewHolder {
            private final TextView nameView;
            private final TextView variationView;
            private final TextView descriptionView;
            private final TextView skuView;
            private final TextView priceView;
            private final TextView quantityView;
            private final ImageButton editButton;
            private final ImageButton deleteButton;
            private final ImageView itemImageView;
            private final ImageView imagePlaceholder;

            ItemViewHolder(@NonNull View itemView) {
                super(itemView);
                nameView = itemView.findViewById(R.id.item_name);
                variationView = itemView.findViewById(R.id.item_variation);
                descriptionView = itemView.findViewById(R.id.item_description);
                skuView = itemView.findViewById(R.id.item_sku);
                priceView = itemView.findViewById(R.id.item_price);
                quantityView = itemView.findViewById(R.id.item_quantity);
                editButton = itemView.findViewById(R.id.edit_item_button);
                deleteButton = itemView.findViewById(R.id.delete_item_button);
                itemImageView = itemView.findViewById(R.id.item_image);
                imagePlaceholder = itemView.findViewById(R.id.item_image_placeholder);
            }

            void bind(Item item) {
                nameView.setText(item.getName());
                
                // Show variation if available
                if (item.getVariationName() != null && !item.getVariationName().isEmpty()) {
                    variationView.setVisibility(View.VISIBLE);
                    variationView.setText(item.getVariationName());
                } else {
                    variationView.setVisibility(View.GONE);
                }
                
                // Show description if available
                if (item.getDescription() != null && !item.getDescription().isEmpty()) {
                    descriptionView.setVisibility(View.VISIBLE);
                    descriptionView.setText(item.getDescription());
                } else {
                    descriptionView.setVisibility(View.GONE);
                }
                
                // Show SKU if available
                if (item.getSku() != null && !item.getSku().isEmpty()) {
                    skuView.setText("SKU: " + item.getSku());
                } else {
                    skuView.setText("");
                }
                
                // Format price with currency symbol
                String currencySymbol = Currency.getInstance(Locale.getDefault()).getSymbol();
                priceView.setText(String.format("%s%.2f", currencySymbol, item.getPrice()));
                
                // Show quantity
                quantityView.setText("In stock: " + item.getQuantity());
                
                // Load item image if available
                if (item.getImagePath() != null) {
                    Bitmap bitmap = itemManager.loadItemImage(item);
                    if (bitmap != null) {
                        itemImageView.setImageBitmap(bitmap);
                        imagePlaceholder.setVisibility(View.GONE);
                    } else {
                        imagePlaceholder.setVisibility(View.VISIBLE);
                    }
                } else {
                    itemImageView.setImageBitmap(null);
                    imagePlaceholder.setVisibility(View.VISIBLE);
                }
                
                // Set up edit button
                editButton.setOnClickListener(v -> {
                    Intent intent = new Intent(ItemListActivity.this, ItemEntryActivity.class);
                    intent.putExtra(ItemEntryActivity.EXTRA_ITEM_ID, item.getId());
                    addItemLauncher.launch(intent);
                });
                
                // Set up delete button
                deleteButton.setOnClickListener(v -> {
                    new AlertDialog.Builder(ItemListActivity.this)
                            .setTitle("Delete Item")
                            .setMessage("Are you sure you want to delete this item?")
                            .setPositiveButton("Delete", (dialog, which) -> {
                                itemManager.removeItem(item.getId());
                                refreshItems();
                                Toast.makeText(ItemListActivity.this, "Item deleted", Toast.LENGTH_SHORT).show();
                            })
                            .setNegativeButton("Cancel", null)
                            .show();
                });
            }
        }
    }
}

File: app/src/main/java/com/electricdreams/numo/feature/items/ItemListActivity.kt
package com.electricdreams.numo.feature.items

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.electricdreams.numo.R
import com.electricdreams.numo.core.model.Item
import com.electricdreams.numo.core.util.ItemManager
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.Collections

class ItemListActivity : AppCompatActivity() {

    private lateinit var itemManager: ItemManager
    private lateinit var recyclerView: RecyclerView
    private lateinit var emptyView: View
    private lateinit var bottomActions: LinearLayout
    private lateinit var fabAddItem: ImageButton
    private lateinit var doneReorderButton: ImageButton
    private lateinit var adapter: ItemAdapter
    private lateinit var itemTouchHelper: ItemTouchHelper

    // Reordering mode state
    private var isReorderingMode = false

    private val addItemLauncher: ActivityResultLauncher<Intent> =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                refreshItems()
                setResult(Activity.RESULT_OK) // Propagate result back
            }
        }

    private val csvPickerLauncher: ActivityResultLauncher<String> =
        registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
            if (uri != null) {
                importCsvFile(uri)
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_item_list)

        // Set up back button
        findViewById<View?>(R.id.back_button)?.setOnClickListener {
            if (isReorderingMode) {
                exitReorderingMode()
            } else {
                finish()
            }
        }

        recyclerView = findViewById(R.id.items_recycler_view)
        emptyView = findViewById(R.id.empty_view)
        bottomActions = findViewById(R.id.bottom_actions)
        fabAddItem = findViewById(R.id.fab_add_item)
        doneReorderButton = findViewById(R.id.done_reorder_button)
        val importCsvButton: Button = findViewById(R.id.import_csv_button)
        val clearItemsButton: TextView = findViewById(R.id.clear_items_button)

        itemManager = ItemManager.getInstance(this)

        recyclerView.layoutManager = LinearLayoutManager(this)
        adapter = ItemAdapter(itemManager.getAllItems())
        recyclerView.adapter = adapter

        // Set up drag-and-drop reordering
        setupDragAndDrop()

        updateEmptyViewVisibility()

        fabAddItem.setOnClickListener {
            val intent = Intent(this, ItemEntryActivity::class.java)
            addItemLauncher.launch(intent)
        }

        doneReorderButton.setOnClickListener {
            exitReorderingMode()
        }

        importCsvButton.setOnClickListener {
            csvPickerLauncher.launch("text/csv")
        }

        clearItemsButton.setOnClickListener {
            showClearAllDialog()
        }
    }

    override fun onResume() {
        super.onResume()
        refreshItems()
    }

    private fun refreshItems() {
        adapter.updateItems(itemManager.getAllItems())
        updateEmptyViewVisibility()
    }

    private fun updateEmptyViewVisibility() {
        val hasItems = adapter.itemCount > 0
        if (hasItems) {
            emptyView.visibility = View.GONE
            recyclerView.visibility = View.VISIBLE
        } else {
            emptyView.visibility = View.VISIBLE
            recyclerView.visibility = View.GONE
        }
    }

    private fun showClearAllDialog() {
        AlertDialog.Builder(this)
            .setTitle(R.string.item_list_dialog_clear_all_title)
            .setMessage(R.string.item_list_dialog_clear_all_message)
            .setPositiveButton(R.string.item_list_dialog_clear_all_positive) { _, _ ->
                itemManager.clearItems()
                refreshItems()
                setResult(Activity.RESULT_OK)
                Toast.makeText(this, getString(R.string.item_list_toast_all_items_cleared), Toast.LENGTH_SHORT).show()
            }
            .setNegativeButton(R.string.common_cancel, null)
            .show()
    }

    private fun importCsvFile(uri: Uri) {
        try {
            val tempFile = File(cacheDir, "import_catalog.csv")

            contentResolver.openInputStream(uri)?.use { inputStream ->
                FileOutputStream(tempFile).use { outputStream ->
                    copyStream(inputStream, outputStream)
                }
            } ?: run {
                Toast.makeText(this, getString(R.string.item_list_toast_failed_open_csv), Toast.LENGTH_SHORT).show()
                return
            }

            val importedCount = itemManager.importItemsFromCsv(tempFile.absolutePath, true)

            if (importedCount > 0) {
                Toast.makeText(this, getString(R.string.item_list_toast_imported_items, importedCount), Toast.LENGTH_SHORT).show()
                refreshItems()
                setResult(Activity.RESULT_OK)
            } else {
                Toast.makeText(this, getString(R.string.item_list_toast_no_items_imported), Toast.LENGTH_SHORT).show()
            }
        } catch (e: IOException) {
            Log.e(TAG, "Error importing CSV file: ${e.message}", e)
            Toast.makeText(this, getString(R.string.item_list_toast_error_importing_csv, e.message), Toast.LENGTH_SHORT).show()
        }
    }

    private fun copyStream(inputStream: InputStream, outputStream: OutputStream) {
        val buffer = ByteArray(4096)
        var bytesRead: Int
        while (true) {
            bytesRead = inputStream.read(buffer)
            if (bytesRead == -1) break
            outputStream.write(buffer, 0, bytesRead)
        }
    }

    private fun setupDragAndDrop() {
        val callback = object : ItemTouchHelper.Callback() {
            override fun getMovementFlags(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder
            ): Int {
                val dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
                return makeMovementFlags(dragFlags, 0)
            }

            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
            ): Boolean {
                val fromPosition = viewHolder.adapterPosition
                val toPosition = target.adapterPosition
                adapter.moveItem(fromPosition, toPosition)
                return true
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                // Not used - swipe disabled
            }

            override fun isLongPressDragEnabled(): Boolean = false

            override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
                super.onSelectedChanged(viewHolder, actionState)
                if (actionState == ItemTouchHelper.ACTION_STATE_DRAG) {
                    viewHolder?.itemView?.alpha = 0.9f
                    viewHolder?.itemView?.elevation = 8f
                }
            }

            override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
                super.clearView(recyclerView, viewHolder)
                viewHolder.itemView.alpha = 1.0f
                viewHolder.itemView.elevation = 0f
                // Persist the new order when drag ends
                adapter.commitReorder()
            }
        }

        itemTouchHelper = ItemTouchHelper(callback)
        itemTouchHelper.attachToRecyclerView(recyclerView)
    }

    fun startDragging(viewHolder: RecyclerView.ViewHolder) {
        itemTouchHelper.startDrag(viewHolder)
    }

    // ----- Reordering Mode -----

    /**
     * Enter reordering mode - shows drag handles and replaces add button with done button.
     */
    fun enterReorderingMode() {
        if (isReorderingMode) return
        isReorderingMode = true

        // Swap buttons with cross-fade animation
        fabAddItem.animate()
            .alpha(0f)
            .setDuration(150)
            .withEndAction {
                fabAddItem.visibility = View.GONE
                doneReorderButton.visibility = View.VISIBLE
                doneReorderButton.alpha = 0f
                doneReorderButton.animate()
                    .alpha(1f)
                    .setDuration(150)
                    .start()
            }
            .start()

        // Hide bottom actions during reordering
        bottomActions.animate()
            .alpha(0f)
            .setDuration(150)
            .withEndAction {
                bottomActions.visibility = View.GONE
            }
            .start()

        // Refresh adapter to show drag handles
        adapter.setReorderingMode(true)
    }

    /**
     * Exit reordering mode - hides drag handles and restores add button.
     */
    private fun exitReorderingMode() {
        if (!isReorderingMode) return
        isReorderingMode = false

        // Swap buttons with cross-fade animation
        doneReorderButton.animate()
            .alpha(0f)
            .setDuration(150)
            .withEndAction {
                doneReorderButton.visibility = View.GONE
                fabAddItem.visibility = View.VISIBLE
                fabAddItem.alpha = 0f
                fabAddItem.animate()
                    .alpha(1f)
                    .setDuration(150)
                    .start()
            }
            .start()

        // Show bottom actions again
        bottomActions.visibility = View.VISIBLE
        bottomActions.alpha = 0f
        bottomActions.animate()
            .alpha(1f)
            .setDuration(150)
            .start()

        // Refresh adapter to hide drag handles
        adapter.setReorderingMode(false)
    }

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        if (isReorderingMode) {
            exitReorderingMode()
        } else {
            super.onBackPressed()
        }
    }

    private inner class ItemAdapter(items: List<Item>) :
        RecyclerView.Adapter<ItemAdapter.ItemViewHolder>() {

        private val itemsList: MutableList<Item> = items.toMutableList()
        private var pendingFromPosition: Int = -1
        private var pendingToPosition: Int = -1
        private var inReorderingMode: Boolean = false

        fun updateItems(newItems: List<Item>) {
            itemsList.clear()
            itemsList.addAll(newItems)
            pendingFromPosition = -1
            pendingToPosition = -1
            notifyDataSetChanged()
        }

        fun setReorderingMode(enabled: Boolean) {
            inReorderingMode = enabled
            notifyDataSetChanged()
        }

        fun moveItem(fromPosition: Int, toPosition: Int) {
            if (fromPosition < 0 || fromPosition >= itemsList.size ||
                toPosition < 0 || toPosition >= itemsList.size) {
                return
            }
            // Track the overall drag operation (from start to end)
            if (pendingFromPosition == -1) {
                pendingFromPosition = fromPosition
            }
            pendingToPosition = toPosition

            // Move item in local list for visual feedback
            Collections.swap(itemsList, fromPosition, toPosition)
            notifyItemMoved(fromPosition, toPosition)
        }

        fun commitReorder() {
            if (pendingFromPosition != -1 && pendingToPosition != -1 &&
                pendingFromPosition != pendingToPosition) {
                // Persist the reorder to storage
                itemManager.reorderItems(pendingFromPosition, pendingToPosition)
                setResult(Activity.RESULT_OK)
            }
            pendingFromPosition = -1
            pendingToPosition = -1
            // Refresh dividers after reorder
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_product, parent, false)
            return ItemViewHolder(view)
        }

        override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
            val item = itemsList[position]
            holder.bind(item, position == itemsList.size - 1, inReorderingMode)
        }

        override fun getItemCount(): Int = itemsList.size

        inner class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            private val nameView: TextView = itemView.findViewById(R.id.item_name)
            private val variationView: TextView = itemView.findViewById(R.id.item_variation)
            private val priceView: TextView = itemView.findViewById(R.id.item_price)
            private val quantityView: TextView = itemView.findViewById(R.id.item_quantity)
            private val itemImageView: ImageView = itemView.findViewById(R.id.item_image)
            private val imagePlaceholder: ImageView = itemView.findViewById(R.id.item_image_placeholder)
            private val divider: View = itemView.findViewById(R.id.divider)
            private val dragHandle: ImageView = itemView.findViewById(R.id.drag_handle)

            @SuppressLint("ClickableViewAccessibility")
            fun bind(item: Item, isLast: Boolean, isReordering: Boolean) {
                // Item name
                nameView.text = item.name ?: ""

                // Variation (grey text)
                if (!item.variationName.isNullOrEmpty()) {
                    variationView.text = item.variationName
                    variationView.visibility = View.VISIBLE
                } else {
                    variationView.visibility = View.GONE
                }

                // Price
                priceView.text = item.getFormattedPrice()

                // Stock quantity (only if tracking inventory)
                if (item.trackInventory) {
                    quantityView.text = "${item.quantity} in stock"
                    quantityView.visibility = View.VISIBLE
                } else {
                    quantityView.visibility = View.GONE
                }

                // Image
                if (!item.imagePath.isNullOrEmpty()) {
                    val bitmap = itemManager.loadItemImage(item)
                    if (bitmap != null) {
                        itemImageView.setImageBitmap(bitmap)
                        imagePlaceholder.visibility = View.GONE
                    } else {
                        itemImageView.setImageBitmap(null)
                        imagePlaceholder.visibility = View.VISIBLE
                    }
                } else {
                    itemImageView.setImageBitmap(null)
                    imagePlaceholder.visibility = View.VISIBLE
                }

                // Hide divider on last item
                divider.visibility = if (isLast) View.GONE else View.VISIBLE

                // Drag handle visibility based on reordering mode
                dragHandle.visibility = if (isReordering) View.VISIBLE else View.GONE

                // Drag handle - start dragging on touch (only when visible)
                if (isReordering) {
                    dragHandle.setOnTouchListener { _, event ->
                        if (event.actionMasked == MotionEvent.ACTION_DOWN) {
                            startDragging(this)
                        }
                        false
                    }
                } else {
                    dragHandle.setOnTouchListener(null)
                }

                // Click behavior depends on mode
                if (isReordering) {
                    // In reordering mode, clicking does nothing (only drag works)
                    itemView.setOnClickListener(null)
                    itemView.setOnLongClickListener(null)
                } else {
                    // Normal mode: click to edit, long-press to enter reordering mode
                    itemView.setOnClickListener {
                        val intent = Intent(this@ItemListActivity, ItemEntryActivity::class.java)
                        intent.putExtra(ItemEntryActivity.EXTRA_ITEM_ID, item.id)
                        addItemLauncher.launch(intent)
                    }

                    itemView.setOnLongClickListener {
                        enterReorderingMode()
                        true
                    }
                }
            }
        }
    }

    companion object {
        private const val TAG = "ItemListActivity"
    }
}
